Web responsive form line breaks

danieltorres7 opened the pull request:
On screens with less than 768px content is limited to 2 columns
On mobile titles have its own line

Depends:
- [x] #513

danieltorres7 committed:
Form view improvements to fit smaller screens

On screens with less than 768px content is limited to 2 columns.

On mobile field labels have their own line.

Fix elements width for XS screens

danieltorres7 commented:
What do you think @lasley @pedrobaeza @leorochael. Web responsive form views should adapt to better fit smaller screens.

danieltorres7 commented:
Before:
===
![captura de tela de 2017-01-02 15-49-54](https://cloud.githubusercontent.com/assets/18170025/21593872/3c5dec80-d103-11e6-9d57-9983ce0526b2.png)

After:
===
![captura de tela de 2017-01-02 15-46-37](https://cloud.githubusercontent.com/assets/18170025/21593880/460eaea4-d103-11e6-8542-dd2107fde755.png)

pedrobaeza commented on web_responsive/static/src/less/form_view.less:
```diff
@@ -0,0 +1,49 @@
+.o_form_view {
+    > .o_form_sheet_bg > .o_form_sheet {
+        max-width: 1150px; // increased form_sheet width for larger screens
```
Will this collide with _web_sheet_full_width_ module?

danieltorres7 replied:
Don't know, will test...

danieltorres7 replied:
I've removed `max-width`, should be OK now...

danieltorres7 commented:
Just added better screenshots above...

pedrobaeza commented:
Down arrows from right part are lost on mobile view:

![seleccion_009](https://cloud.githubusercontent.com/assets/7165771/21593947/3d9da4fe-d11d-11e6-842f-c3ca70c27ce9.png)

danieltorres7 commented:
@pedrobaeza, just tested on last commit runbot: http://3276840-514-f788ea.runbot2.odoo-community.org/?db=3276840-514-f788ea-all and got these screenshots (both with down arrows):
![captura de tela de 2017-01-02 16-08-09](https://cloud.githubusercontent.com/assets/18170025/21594120/ecf3bca8-d105-11e6-8716-f53c7739041d.png)
![captura de tela de 2017-01-02 16-07-54](https://cloud.githubusercontent.com/assets/18170025/21594121/ef03290c-d105-11e6-872d-c8d710df5654.png)

danieltorres7 commented:
@pedrobaeza, where did you see the arrows dissapearing? I'm trying to reproduce that...

pedrobaeza commented:
I'm using Firefox. That should be the problem. I'm also not seeing that cool buttons for editing or removing the image.

pedrobaeza commented:
OK, buttons appear at a very low resolution, but that's not useful I think.

danieltorres7 commented:
> OK, buttons appear at a very low resolution, but that's not useful I think.

This is default Odoo behaviour, I'm not changing that...

pedrobaeza commented:
OK, I didn't know, but maybe we can change class for showing it at a higher resolution, as mobile views are greater than that I think.

danieltorres7 commented:
> OK, I didn't know, but maybe we can change class for showing it at a higher resolution, as mobile views are greater than that I think.

OK, I'll consider that for another PR. 

BTW, I would prefer that this PR is merged only after rebasing onto #513

lasley commented:
> BTW, I would prefer that this PR is merged only after rebasing onto #513

To mark that a PR is dependent on another, put this in your original comment with a check list. Something like:

```
Depends:
- [ ] #513
```

Then once it's merged and you rebase, you can just check it off the list 🚀

leorochael commented:
> OK, I didn't know, but maybe we can change class for showing it at a higher resolution, as mobile views are greater than that I think.

@pedrobaeza: You're right about mobile devices having higher resolution, and the resolution will only get higher with time.

The problem is that it can be argued that what Odoo does here is wrong: it uses screen resolution as a proxy for the absence of a mouse, as it would require having a mouse to hover over the image before seeing the edit icons in high resolution.

In discussions of the design community ([like this](https://css-tricks.com/forums/topic/conditional-css-for-touch-screen-devices/)), the consensus seems to be that hover events should only ever be used to add optional embellishments to the design, but not be necessary for the functionality, no matter the resolution.

So I guess we could have `web_responsive` *fix* Odoo by making the buttons always be present.

But like @danieltorres7 said, this should go into another PR.

danieltorres7 committed:
Form image editing controls always present

Make image editing controls always available, instead of depending
on resolution or hover.

danieltorres7 commented:
Since image editing controls are related to forms I just committed an addition to make them always visible.

pedrobaeza commented:
@Yajo, please review

yajo commented:
Seems a good IMP. I'll mark as WIP until the other PR is merged and then review 😉

leorochael commented:
@Yajo, the behaviour of this PR is not really affected by the presence of #513. It just looks prettier when the menus are also better adapted to mobile.

You can review it at your convenience, regardless of #513.

yajo commented on web_responsive/static/src/less/form_view.less:
```diff
@@ -0,0 +1,66 @@
+                            padding: 0;
+                            display: block;
+                            padding: 0;
+                            width: auto!important; // important is to override width fixed by element style on base
```
Please replace `!important`s for more concise CSS selectors.

danieltorres7 replied:
This import is to override the base template where td's have width: 100%!important. Precise selector can't do the job.

yajo commented on web_responsive/static/src/less/form_view.less:
```diff
@@ -0,0 +1,66 @@
+                            max-width: 100%;
+                            &.o_td_label {
+                                border-right: 0;
+                                width: 94%!important; // keep 6% space on line to fit checkboxes
```
Replace.

danieltorres7 replied:
This important is to override the previous important on the same less file.

yajo replied:
Same case as https://github.com/OCA/web/pull/513#discussion_r94752243

leorochael replied:
Like in #513, there's a `style="width: 100%"` attribute directly on the element (`<td>` in this case) I don't know how to make a selector more specific than the `style` attribute.

We haven't figured out yet where this `style` attribute on the `<td>` is coming from.

Does anyone have any idea?

leorochael replied:
Found [the culprit](https://github.com/odoo/odoo/blob/3d085f632d4c5a8aece92731494e1dbbc8923049/addons/web/static/src/js/views/form_view.js#L1499):

     $td.css("width", (nonlabel_width*parseInt($td.attr('colspan')))+"%");

@Yajo what do you suggest?

yajo replied:
Hmm a tricky one... What about this?

1. Make sure your solution does not break when there are colspans.
2. If it works fine, then I suggest you put an explanation above this line, seems like it would be hard to fix this upstream, and after all they're gonna open the responsive client for v11.

yajo commented on web_responsive/static/src/less/form_view.less:
```diff
@@ -0,0 +1,66 @@
+            text-align: center;
+        }
+        > .fa.o_select_file_button {
+            background: #7c7bad;
```
Use color variables from where you now know 😉

danieltorres7 replied:
Changing...

yajo commented on web_responsive/static/src/less/form_view.less:
```diff
@@ -0,0 +1,66 @@
+            background: #7c7bad;
+        }
+        > .fa.o_clear_file_button {
+            background: #d9534f;
```
Color variables.

danieltorres7 replied:
Changing...

yajo reviewed (changes requested):
This seems much better than it was indeed. However it seems incomplete.

See, this is 458×575px window (FF):

![captura el 2017-01-04 a las 10 02 23](https://cloud.githubusercontent.com/assets/973709/21636925/e6bfe0cc-d264-11e6-81e8-89f7d87bd444.png)

Do you see the horizontal scroll bar? See what happens when I scroll to the right:

![captura el 2017-01-04 a las 10 03 07](https://cloud.githubusercontent.com/assets/973709/21636946/0028592c-d265-11e6-8b15-c7d046dede57.png)

I think the problem is in the chatter widget. it gets wider than its parent container. Can you check that please?

danieltorres7 commented:
> I think the problem is in the chatter widget. it gets wider than its parent container. Can you check that please?

Your right, #513 should fix that after merging.

mileo commented:
@lasley "Other than that though, general rule is one final commit per PR - unless multiple contributors or migrations" Is this recorded in any document written in the OCA? (sorry for the off topic)

lasley commented:
@mileo [Contribution Docs, Commit guidelines](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#git). "One logical changeset" typically works out to one per PR, but obviously can vary depending on context - such as multiple authors.

pedrobaeza commented:
What it still not reflected in guidelines is the agreement we made between several regular contributors for the reviewing process: don't squash commits until the PR is approved. This way, if you have reviewed full PR and make comments, you only need to review again the new commits to see if your comments have been attended. Once the review is done and you approve the PR, that new extra commits like "PEP8", "Requested changes" and so on must be squashed before merging. If all the commits come from the same person and it's only one logical changeset, this can be done using the GitHub feature "Squash and merge". If not, any contributor with write access on the repository can do it (if enabled "Allow to edit to maintainers" on the PR), or the author of the PR otherwise.

This can be included in the guidelines for being clearer if any volunteer.

moylop260 commented:
The best comment about squash commit @pedrobaeza

I can help but my bad english make many typo feedback and you finish writing all again haha

What about create your PR to guidelines?

pedrobaeza commented:
OK, I'll do it sometime in the day.

danieltorres7 commented:
@Yajo, requested changes and final fixes applied. We are ready for the final review.

pedrobaeza commented:
Dependency merged. Should you rebase?

yajo reviewed (changes requested):
Rebase, I want to check if nothing breaks now that #513 is merged.

leorochael commented:
@danieltorres7 is till the end of the week. If he adds me as a developer to his fork I can rebase for him.

pedrobaeza commented:
I have rebased the branch.

@Yajo, OCA/web maintainers can edit PR branches.

@leorochael, what we do when working collaboratively is to push our PR branches to the team, so anyone in the team can modify them. It's only a suggestion, obviously :smile:

yajo reviewed (changes requested):
As you can see below in a screenshot of a partner form, there's still the `.oe_chatter` problem. It has a `min-width` that must be removed for it to not break the layout:

![captura de pantalla de 2017-01-17 09-14-38](https://cloud.githubusercontent.com/assets/973709/22012692/b8cfa2a2-dc95-11e6-8926-dd5250dcc3fa.png)

yajo committed:
Fix problem with overflowing chatter widget.

pedrobaeza commented:
Please squash the first 4 commits, Daniel.

danieltorres7 commented:
Squashed last 4 commits into 2 commits by relevant changes on each.

Ready for merging in my opinion.

rafaelbn reviewed (approved):
Thanks!
